lib/Conversion/TosaToXTenNN.cpp

```diff
@@ -0,0 +1,225 @@
+///\return true represents a quantize cast
+///\return false does not represent a quantize cast
+bool isF32ToI8Cast(tosa::CastOp castOp) {
+  TensorType inputTensorType =
+      cast<TensorType>(castOp->getOperand(0).getType());
+  TensorType outputTensorType =
+      cast<TensorType>(castOp->getResult(0).getType());
+  return inputTensorType.getElementType().isF32() &&
+         outputTensorType.getElementType().isInteger(8) &&
+         outputTensorType.getElementType().isSignlessInteger();
+}
+
+/// Get the Log2 \p value of the float. If the \p value is not an exact
+/// power-of-two return none.
+///
+///\param value floating point value we are checking
+///\return std::optional<int32_t> none if \p value is not a power of two and an
+/// integer if it is.
+std::optional<int32_t> getLog2Value(float value) {
+  float log2Value = std::log2(value);
+  float integerPart = 0.0;
+  float fractionalPart = std::modf(log2Value, &integerPart);
+  if (fractionalPart != 0.0) {
+    return std::nullopt;
+  }
+  return (int32_t)integerPart;
+}
+
+/// Checks to see that the two consecutive casts can be represented by quantize
+/// and dequantize operations.
+class CastsToQDQOps : public OpRewritePattern<tosa::CastOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::CastOp castOp,
+                                PatternRewriter &rewriter) const override {
+    auto quantizeOp =
+        dyn_cast_or_null<tosa::CastOp>(castOp->getOperand(0).getDefiningOp());
+    if (!quantizeOp) {
+      return rewriter.notifyMatchFailure(castOp->getOperand(0).getLoc(),
+                                         "expected cast -> cast pattern.");
+    }
+
+    // Dequantize is from i8 -> f32 here we need to check for that
+    if (!isI8ToF32Cast(castOp)) {
+      return rewriter.notifyMatchFailure(
+          castOp->getLoc(), "dequantize operation input and output are not i8 "
+                            "and f32 respectively.");
+    }
+
+    // Quantize is from f32 -> i8 here we need to check for that
+    if (!isF32ToI8Cast(quantizeOp)) {
+      return rewriter.notifyMatchFailure(castOp->getLoc(),
+                                         "quantize operation input and output "
+                                         "are not f32 and i8 respectively.");
+    }
+
+    auto outputTensorType =
+        cast<TensorType>(quantizeOp->getOperand(0).getType());
+    Type newOutputType = outputTensorType.cloneWith(
+        {}, IntegerType::get(rewriter.getContext(), 8,
+                             IntegerType::SignednessSemantics::Signed));
+
+    auto newQuantizeOp = rewriter.create<amd::xten_nn::QuantizeOp>(
+        quantizeOp->getLoc(), newOutputType, quantizeOp->getOperand(0),
+        /*shift*/ (int32_t)0);
+    rewriter.replaceOpWithNewOp<amd::xten_nn::DequantizeOp>(
+        castOp, outputTensorType, newQuantizeOp->getResult(0),
+        /*shift*/ (int32_t)0);
+    rewriter.eraseOp(quantizeOp);
```

Comment:
This is only safe when quantizeOp had a single user. Can you add a test case where that ops has two users and then fix the code here by checking num_users?